### PR TITLE
QA Training and New Architecture

### PR DESCRIPTION
**Pull request by EricFillion**
Introduces QA Training and an all-new architecture.  Please refer to the diagram and the messages I sent you over private messages for an overview of the new architecture. I'm happy to answer any questions or provide more detail if requested. 

**Testing:** 
I've included two basic tests within test_qa_training.py.

We will expand these tests to read the outputted CSV files from QATrainer.eval() and QATrainer.test() for more thorough testing. 

We will also use a larger dataset. However, we have to be careful not to commit the dataset directly to the repository. Instead, we can set up a system where the dataset is downloaded from the internet upon testing and is saved for future use. 

I've also updated the tests within test_qa_multi.py and test_qa_util.py to reflect the new architecture. 

**Breaking changes:** 
Before using answer_to_questions or answer_question, you must call init_qa(). 

**Steps before being merged into master:** 
- clean up 
- add comments
- complete QATrainer.test()
- improve progress indicator for training 

**Requested feedback:** 
Do you agree with the new proposed architecture? 
Any feedback on naming public functions/parameter names?  
Are there any features you think should be added or removed?

Feel free to commit straight to this PR for obvious stuff.

### CODE REVIEW

File: happytransformer/happy_bert.py [hunk 1]
```diff
@@ -39,15 +39,25 @@ class HappyBERT(HappyTransformer):
...
-        self.qa = None   # Question Answering
-        self.tokenizer = BertTokenizer.from_pretrained(model)
+
+        #todo separate tokenizer for each model
+        self.tokenizer = BertTokenizerFast.from_pretrained("bert-base-uncased")
         self.masked_token = self.tokenizer.mask_token
         self.sep_token = self.tokenizer.sep_token
         self.cls_token = self.tokenizer.cls_token
 
+        # ------------------------ QA
```
Comment (ted537) [inline happytransformer/happy_bert.py:53]: The slew of `__qa_` variables seems weird. Holding a single class instance might be cleaner. Also, it would be more maintainable if this top level logic was stored in `happy_transformer.py`, with implementation specifics in `happy_bert.py`.

File: happytransformer/happy_bert.py [hunk 2]
```diff
@@ -131,63 +136,72 @@ def __is_one_sentence(self, text):
                     sentence_found = True
                     break
         return True
+#-------------------------------------------------------#
+
+                # QUESTION ANSWERING #
+#-------------------------------------------------------#
+
+    def init_qa(self, model='bert-large-uncased-whole-word-masking-finetuned-squad'):
```
Comment (ted537) [inline happytransformer/happy_bert.py:144]: This is a weird method. The only way to access it is through an instance of HappyBERT. Let's say I run the following.

```
happy_bert = HappyBERT()
happy_bert.init_qa()
```

Now the models and tokenizers for 'bert-base-uncased' and 'bert-large-uncased-whole-word-masking-finetuned-squad' are being stored in memory. This prohibitively and unecessarily increases the memory requirements of HappyTransformer.

File: happytransformer/happy_bert.py [hunk 3]
```diff
@@ -131,63 +136,72 @@ def __is_one_sentence(self, text):
...
+        self.__qa_model.eval()
+
+        if self.gpu_support == 'cuda':
+            self.__qa_model.to('cuda')
+
+        self.__qa_runner = AnswerQuestionRunner(self._model_name, self.__qa_model, self.__qa_tokenizer)
+        self.__qa_init = True
+
+    def answers_to_question(self, question, context, k=3):
+        if self.__qa_init:
```
Comment (ted537) [inline happytransformer/happy_bert.py:160]: It seems weird that MLM and NSP models are loaded by default, but QA models need to have an initialization method

File: happytransformer/runners/runner_answer_question.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
+import torch
+
+from collections import namedtuple
+
+
+class AnswerQuestionRunner():
+    SumPair = namedtuple('SumPair', ['idx1', 'idx2', 'sum'])
```
Comment (ted537) [inline happytransformer/runners/runner_answer_question.py:7]: Not sure why this type is stored within the class

File: tests/test_qa_util.py [hunk 5]
```diff
@@ -2,22 +2,23 @@
...
-        SumPair(idx1=1, idx2=2, sum=8),  # 4+4
-        SumPair(idx1=2, idx2=2, sum=7)  # 3+4
+        qa_runner.SumPair(idx1=0, idx2=0, sum=14),  # 7+7
+        qa_runner.SumPair(idx1=0, idx2=1, sum=13),  # 7+6
+        qa_runner.SumPair(idx1=0, idx2=2, sum=11),  # 7+4
+        qa_runner.SumPair(idx1=1, idx2=2, sum=8),  # 4+4
+        qa_runner.SumPair(idx1=2, idx2=2, sum=7)  # 3+4
     ]
-    computed_pairs = biggest_sums(items_a, items_b)
+    computed_pairs = qa_runner.biggest_sums(items_a, items_b)
```
Comment (ted537) [inline tests/test_qa_util.py:21]: This method should be a standalone function; it has no inputs other than items_a and items_b

File: happytransformer/trainers/trainer_qa.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+            self._output_result_to_csv(output_filepath, fieldnames, results)
+
+        return score
+
+
+    def test(self, filepath, args, output_filepath):
+        #todo
+        pass
+
+    def __get_train_eval_data(self, filepath):
```
Comment (ted537) [inline happytransformer/trainers/trainer_qa.py:125]: doesn't use self; can be standalone

File: happytransformer/trainers/trainer_qa.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+                answer = {}
+                answer["answer_text"] = row['answer_text']
+                answer["answer_start"] = int(row['answer_start'])
+                answers.append(answer)
+        csv_file.close()
+
+        return contexts, questions, answers
+
+
+    def __add_end_idx(self, contexts, answers):
```
Comment (ted537) [inline happytransformer/trainers/trainer_qa.py:149]: doesn't use self; can be standalone

File: happytransformer/trainers/trainer_qa.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        start_positions = []
+        end_positions = []
+        for i in range(len(answers)):
+            start_positions.append(encodings.char_to_token(i, answers[i]['answer_start']))
+            end_positions.append(encodings.char_to_token(i, answers[i]['answer_end'] - 1))
+            # if None, the answer passage has been truncated
+            if start_positions[-1] is None:
+                start_positions[-1] = self.tokenizer.model_max_length
+            if end_positions[-1] is None:
+                end_positions[-1] = self.tokenizer.model_max_length
```
Comment (ted537) [inline happytransformer/trainers/trainer_qa.py:179]: Retroactively fixing up the positions seems odd.

**Review by ted537 — CHANGES_REQUESTED** [review]

File: happytransformer/trainer.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+    def _output_result_to_csv(self, output_filepath, fieldnames, results):
+            with open(output_filepath, 'w') as csv_file:
+                csv_writer = DictWriter(csv_file, fieldnames=fieldnames)
+                csv_writer.writeheader()
+                for result in results:
+                    csv_writer.writerow(
+                        result
+                    )
+            csv_file.close()
```
Comment (ted537) [inline happytransformer/trainer.py:93]: csv_file is already closed by the `with` statement. This line is unecessary

File: happytransformer/trainers/trainer_qa.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+            if context[start_idx:end_idx] == gold_text:
+                answer['answer_end'] = end_idx
+            elif context[start_idx - 1:end_idx - 1] == gold_text:
+                answer['answer_start'] = start_idx - 1
+                answer['answer_end'] = end_idx - 1  # When the gold label is off by one character
+            elif context[start_idx - 2:end_idx - 2] == gold_text:
+                answer['answer_start'] = start_idx - 2
+                answer['answer_end'] = end_idx - 2  # When the gold label is off by two characters
+            else:
+                print("error: implement skipping training answer")
```
Comment (ted537) [inline happytransformer/trainers/trainer_qa.py:167]: could you link a SQUAD example where this edge case is necessary?

File: happytransformer/happy_bert.py [hunk 11]
```diff
@@ -39,15 +39,25 @@ class HappyBERT(HappyTransformer):
...
     def __init__(self, model='bert-base-uncased'):
+        # todo remove model parameter. Each model will have its own
         super().__init__(model, "BERT")
         self.mlm = None  # Masked Language Model
         self.nsp = None  # Next Sentence Prediction
-        self.qa = None   # Question Answering
-        self.tokenizer = BertTokenizer.from_pretrained(model)
+
+        #todo separate tokenizer for each model
+        self.tokenizer = BertTokenizerFast.from_pretrained("bert-base-uncased")
```
[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/happy_bert.py:48]: In version 2.0 we'll use a custom tokenizer for each task. When the user calls init_task(model), where "task" may be "qa", "mwp" etc, then a tokenizer is loaded with the given model for the task. 

So, it doesn't matter too much what this is for now, but I'll change it back to model for the time being.

File: happytransformer/runners/runner_answer_question.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
+import torch
+
+from collections import namedtuple
+
+
+class AnswerQuestionRunner():
```
[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/runners/runner_answer_question.py:6]: I agree

File: happytransformer/runners/runner_answer_question.py [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+    SumPair = namedtuple('SumPair', ['idx1', 'idx2', 'sum'])
+
+    def __init__(self, type, model, tokenizer):
+        self.type = type # BERT, ROBERTA, ALBERT etc
+        self.model = model
+        self.tokenizer = tokenizer
+
+    def run_answers_to_question(self,  question, context, k=3):
```
[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/runners/runner_answer_question.py:14]: To distinguish it from answers_to_questions method within HappyBERT. I'm open to changing it tho.

File: happytransformer/runners/runner_answer_question.py [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+            0 if idx <= sep_id_index else 1
+            for idx, _ in enumerate(input_ids)
+        ]
+        with torch.no_grad():
+            return self.model(
+                input_ids=torch.tensor([input_ids]),
+                token_type_ids=torch.tensor([before_after_ids])
+            )
+
+    def biggest_sums(self, items_a, items_b):
```
[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/runners/runner_answer_question.py:73]: I agree, it will likely be used by other runners, I moved it to a file called runner_util.py

File: happytransformer/runners/runner_answer_question.py (hunk 4 above)
Comment (EricFillion) [reply happytransformer/runners/runner_answer_question.py:7]: Moved to runner_util.py

File: tests/test_qa_util.py (hunk 5 above)
Comment (EricFillion) [reply tests/test_qa_util.py:21]: Moved to runner_util.py

File: happytransformer/trainers/trainer_qa.py (hunk 6 above)
Comment (EricFillion) [reply happytransformer/trainers/trainer_qa.py:125]: I agree

File: happytransformer/trainers/trainer_qa.py (hunk 7 above)
Comment (EricFillion) [reply happytransformer/trainers/trainer_qa.py:149]: I agree

File: happytransformer/trainers/trainer_qa.py (hunk 8 above)
Comment (EricFillion) [reply happytransformer/trainers/trainer_qa.py:179]: I agree this seems a little odd. I followed the official documentation. In the future we can adjust it so hopefully we don't have to retroactively fix the positions. But since it's within the official documentation, it likely works.

File: happytransformer/trainer.py (hunk 9 above)
Comment (EricFillion) [reply happytransformer/trainer.py:93]: Nice catch

File: happytransformer/trainers/trainer_qa.py (hunk 10 above)
Comment (EricFillion) [reply happytransformer/trainers/trainer_qa.py:167]: The official documentation does not indicate which examples cause this edge case. It would be a good experiment tho.

File: happytransformer/runners/runner_answer_question.py [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+        """
+        Using the given text, find the answer to the given question and return it.
+
+        :param question: The question to be answered
+        :param text: The text containing the answer to the question
+        :return: The answer to the given question, as a string
+        """
+        return self.run_answers_to_question(question, text,  1)[0]["text"]
+
+    def _tokenize_qa(self, question, context):
```
[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/runners/runner_answer_question.py:48]: We'll discuss modifying #161 to conform to the architecture after this PR is merged. If we merge #161 now, it will lead to merge conflicts.

File: happytransformer/trainers/trainer_qa.py (hunk 10 above)
Comment (ted537) [reply happytransformer/trainers/trainer_qa.py:169]: "sometimes squad answers are off by a character or two" is there a source for this?

File: happytransformer/runners/runner_answer_question.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (ted537) [reply happytransformer/runners/runner_answer_question.py:12]: I don't think the naming difference is necessary to distinguish them. Both methods live on objects anyway.

```
# in happy_transformer.py
def answers_to_question(...):
  # ...
  return self.qa.answers_to_question(...)
```

File: happytransformer/trainers/trainer_qa.py (hunk 10 above)
Comment (EricFillion) [reply happytransformer/trainers/trainer_qa.py:169]: Source: https://huggingface.co/transformers/custom_datasets.html#question-answering-with-squad-2-0

Screen Shot 2021-01-02 at 12 28 29 AM

File: happytransformer/trainer.py [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+        self.model = model
+        self.model_type = model_type
+        self.tokenizer = tokenizer
+        self.device = device
+        self.runner = runner
+        self.logger = logger
+
+
+    def train(self, filepath, args):
+        pass
```
[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/trainer.py:18]: I agree

[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/trainer.py:18]: I also added it for test() and eval()

File: happytransformer/trainer.py [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+
+
+        :param count:
+        :return:
+        """
+
+        x = count / 50
+        order = math.floor(math.log(x, 10))
+
+        update_interval = order ** 10
```
[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/trainer.py:63]: oops, nice catch

File: happytransformer/happy_transformer.py [hunk 18]
```diff
@@ -63,6 +62,9 @@ def __init__(self, model, model_name):
             else "cpu"
         )
 
+        self.device = True if torch.cuda.is_available() else False
```
[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/happy_transformer.py:65]: yeah, we can just stick to use the gpu_support support variable. Maybe rename it to device though.

File: happytransformer/happy_bert.py [hunk 19]
```diff
@@ -131,63 +136,72 @@ def __is_one_sentence(self, text):
...
-            for answer in probabilities
-        ]
\ No newline at end of file
+        if self.__qa_init:
+            return self.__qa_runner.run_answer_question(question, text)
+        else:
+            self._init_model_first_warning("question answering", "init_qa(model_name)")
+
+    def train_qa(self, filepath, args=None):
+        if self.__qa_init:
```
[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/happy_bert.py:181]: I just cleaned this up. It's much better now.

File: happytransformer/trainers/trainer_qa.py [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+We prioritized following the official documentation as close as possible to ensure we're using
+robust methods. And also, to improve maintainability as they update the documentation.
+
+https://huggingface.co/transformers/custom_datasets.html#question-answering-with-squad-2-0
+"""
+
+import torch
+from happytransformer.trainer import Trainer
+import csv
+from torch.utils.data import DataLoader
```
[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/trainers/trainer_qa.py:14]: fixed

File: happytransformer/happy_bert.py (hunk 1 above)
Comment (EricFillion) [reply happytransformer/happy_bert.py:53]: I agree, that does seem cleaner. I did it this way because not every Happy class will have QA functionality, but on second thought you're right. Keeping them private and as None does no harm even if the class does not use them.

File: happytransformer/happy_bert.py (hunk 3 above)
Comment (EricFillion) [reply happytransformer/happy_bert.py:160]: In version 2.0, every task will follow the structure proposed in this PR. This will result in breaking changes, but it's worth it in the long run.

File: happytransformer/runners/runner_answer_question.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/runners/runner_answer_question.py:12]: Okay sounds good. I changed it.

File: happytransformer/trainers/trainer_qa.py (hunk 10 above)
Comment (ted537) [reply happytransformer/trainers/trainer_qa.py:168]: https://gist.github.com/ted537/29a656d7bbce7247a3c89b7d4c011015

I created a script to test for this issue and it didn't come up. I believe it only occurs if you save the SQUAD JSON with the wrong encoding. This occurs because converting to ASCII can reduce certain UTF8 characters away. We can probably add a warning to users in the case where we have to clean up the indices.

TLDR: The SQUAD dataset is correct, but becomes incorrect if saved in the wrong encoding (user error quite possible)

**Comment by ted537** [comment]
# Naming

Could you explain what runner is supposed to signify? I'm having trouble understanding. My first guess was that its an object that does (runs) something, but I'm not seeing the distinction between this and any other function.

# The Init() Problem

Transformer models are large. As a result, loading several models into memory should be avoided when possible. If users run MLM functions, NSP functions, and then QA functions, memory usage will increase significantly, possibly exceeding the limits of their computer. The fact that memory usage permanently increases when a new function is called, and then cannot be redeemed without destroying the entire transformer object or sneakily setting `happy.mlm=None` is not intuitive.

# Structure

If we are restructuring things anyway, then I think now is a good time to consider what a HappyTransformer object is. At present, it appears to be 3 very loosely related pipelines of components. That is there a separate pipeline of tokenizers / models / special rules for each of the transformer models HappyTransformer supports. Given that there is little overlap between QA, MLM, and NSP functions other than possibly sharing a tokenizer, I believe it would make sense to separate these processes completely. While it is nice in theory to have a single top level object, the resource allocation concerns introduce significant caveats.

```
class MaskedPredictor:
  def __init__(self, tokenizer, mlm):
    self.tokenizer = tokenizer
    self.mlm = mlm

  def predict_masks(self,...):
    # ...
    pass
```

**Comment by EricFillion** [comment]
Great points. Let's hop on a zoom call to discuss this briefly. I just need to clarify a few things before I start modifying the code.

**Comment by EricFillion** [comment]
Will be creating a PR into a branch called "v2" so that the master branch doesn't look incomplete as we finish v2.0.0
